drivers/include/periph/gpio_ll.h

```diff
@@ -0,0 +1,498 @@
+ *              nor RAM, nor ROM). Hence, this function will only cost you
+ *              if your platform really requires preparation of any sorts.
+ *
+ * See @ref gpio_ll_write on how to use this function
+ */
+static inline uword_t gpio_ll_prepare_write(gpio_port_t port, uword_t value);
+
+/**
+ * @brief   Perform an `reg = mask` operation on the I/O register of the port
+ *
+ * @warning Some platforms multiplex the "write" I/O register with additional
+ *          functions e.g. for input pin configuration. To prevent unintentional
+ *          side effects prepare a value using @ref gpio_ll_prepare_write that
+ *          will set the bits of non-output pins as needed to not have side
+ *          effects on the state the GPIO port had when calling
+ *          @ref gpio_ll_prepare_write .
+ *
+ * @param[in] port      port to modify
+ * @param[in] value     new state of the GPIO port to write
+ *
+ * Usage:
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.c}
+ *  // emit a square wave from two pins with a phase shift of pi between the
+ *  // waves, e.g.:
+ *  // pin1: _⎍_⎍_⎍_⎍_⎍_⎍_⎍_⎍_⎍
+ *  // pin2: ⎍_⎍_⎍_⎍_⎍_⎍_⎍_⎍_⎍_
+ *  void square_wave(gpio_port_t port, uint8_t pin1, uint8_t pin2)
+ *  {
+ *      uword_t state1 = gpio_ll_prepare_write(port, 1 << pin1);
+ *      uword_t state2 = gpio_ll_prepare_write(port, 1 << pin2);
+ *      while (1) {
+ *          gpio_ll_write(port, state1);
+ *          gpio_ll_write(port, state2);
+ *      }
+ *  }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * @note    If the configuration of the used port changes between the calls
+ *          of @ref gpio_ll_prepare_write and @ref gpio_ll_write the write will
+ *          still have undesired side effects on the configuration of input pins
+ *          if the platform multiplexes the write register with configuration
+ *          for input pins. It is expected that the user of the API either
+ *          exclusively uses a GPIO port or synchronizes with other users to
+ *          update the prepared value on configuration changes.
+ */
+static inline void gpio_ll_write(gpio_port_t port, uword_t value);
+
+/**
+ * @brief   Extract the `gpio_port_t` from a `gpio_t`
+ */
+static inline gpio_port_t gpio_get_port(gpio_t pin);
+
+/**
+ * @brief   Extract the pin number from a `gpio_t`
+ */
+static inline uint8_t gpio_get_pin_num(gpio_t pin);
+
+/**
+ * @brief   Pack a pointer into a @ref gpio_port_t
+ *
+ * @pre     The address in @p addr is not `NULL` ***and*** points to a regular
+ *          memory region (e.g. anywhere in `.data`, `.rodata`, `.bss` or an
+ *          address returned by `malloc()` and friends)
+ *
+ * @return  A value that is distinguishable from any valid port reference and
+ *          from which @p addr can be reconstructed by calling
+ *          @ref gpio_port_unpack_addr
+ */
+static inline gpio_port_t gpio_port_pack_addr(void *addr);
```

Comment:
From the description I don't quite get why this function is needed / when it should be used.

Reply:
Let's say I have a

```C
typedef struct {
    gpio_port_t port;
    uint8_t pin;
} gpio_t;
```

And that `port` may either be a peripheral GPIO port, or a pointer to a device descriptor of an GPIO pin extender. This function it the reverse of it function would allow this multiplexing. Usually, one can just take the address as is and later check if the address is in the peripheral address space (--> peripheral GPIO port), or note (--> `void *` packed into `gpio_port_t`). But since the peripheral address space is platform specific, I added this to the HAL.

I'm not sure that this is actually the way to go for a high level pin based API on top. One could safe some RAM by going for something like 

``` C
typedef union {
    uint16_t bits;
    struct fields {
        uint16_t pin : 5;
        uint16_t port : 5;
        uint16_t device : 6;
    }
} gpio_t;
```

instead and convert the port number to a `gpio_port_t` on demand. But since this is trivial to implement, easy to test, and potentially useful for a high level API, I thought we could just add it.